### PR TITLE
[JUJU-3897] Implement initial cloud domain service and state

### PR DESCRIPTION
**Pull request by wallyworld**
This PR provides an initial cloud domain state and service implementation, on top of the previously committed DDL for clouds, regions, auth types, ca certs etc. 

Note the DDL currently does not have cloud config or region config so these will be done later.

The service methods included here are:
- Upsert
- ListAll
- Get

These will be sufficient to wire up a few of the cloud CLI commands once the facade plumbing is done.

The structs for the db entities are define in `types.go`. These are not all currently used, but are there for when sqlair is introduced. 

## QA steps

run the unit tests

### CODE REVIEW

File: domain/cloud/state/state.go [hunk 1]
```diff
@@ -0,0 +1,441 @@
...
+	cld := &Cloud{
+		ID:               cloudUUID,
+		Name:             cloud.Name,
+		Endpoint:         cloud.Endpoint,
+		IdentityEndpoint: cloud.IdentityEndpoint,
+		StorageEndpoint:  cloud.StorageEndpoint,
+		SkipTLSVerify:    cloud.SkipTLSVerify,
+	}
+
+	row := tx.QueryRow("SELECT id FROM cloud_type WHERE type = ?", cloud.Type)
```
Comment (nvinuesa) [inline domain/cloud/state/state.go:432]: should we `LIMIT 1` here? Otherwise at least we should check that only one row is returned right?

Comment (nvinuesa) [reply domain/cloud/state/state.go:432]: Ok sorry, actually `QueryRow` only keeps the first row and discards the rest, so no need for LIMIT.

File: domain/cloud/state/state.go [hunk 2]
```diff
@@ -0,0 +1,441 @@
...
+		}
+
+		return nil
+	})
+
+	return errors.Trace(err)
+}
+
+func upsertCloud(ctx context.Context, tx *sql.Tx, cloudUUID string, cloud cloud.Cloud) error {
+	dbCloud, err := dbCloudFromCloud(tx, cloudUUID, cloud)
```
Comment (manadart) [inline domain/cloud/state/state.go:279]: Do a join in the main query. See the lease store for an example.

**Review by manadart — COMMENTED** [review]
I don't think we're sufficiently embracing the relational algebra here.

File: domain/cloud/state/state.go (hunk 2 above)
Comment (wallyworld) [reply domain/cloud/state/state.go:279]: If you are talking about what is done in ClaimLease() then that doesn't appear to correctly return an error if the lease type is invalid, like we do here if a cloud with an unknown type is saved.

File: domain/cloud/service/service.go [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+	st State
+}
+
+// NewService returns a new service reference wrapping the input state.
+func NewService(st State) *Service {
+	return &Service{st}
+}
+
+// Upsert inserts or updates the specified cloud.
+func (s *Service) Upsert(ctx context.Context, cloud cloud.Cloud) error {
```
[the remark this answers is not in the document]
Comment (wallyworld) [reply domain/cloud/service/service.go:35]: I thought about separate Create() and Update() methods but IMO unless you really want to see an AlreadyExists error or a NotFound error for the separate methods, a single create/update method seems nicer to use.

File: domain/cloud/state/state.go [hunk 4]
```diff
@@ -0,0 +1,441 @@
...
+		return nil, errors.Trace(err)
+	}
+
+	var uuids []string
+	for uuid := range clouds {
+		uuids = append(uuids, uuid)
+	}
+
+	// Add in the ca certs and regions.
+	caCerts, err := loadCACerts(tx, uuids)
```
[the remark this answers is not in the document]
Comment (wallyworld) [reply domain/cloud/state/state.go:143]: also, this strategy is commonly used to allow for lazy loading of 1-many relations

File: domain/cloud/service/service.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (wallyworld) [reply domain/cloud/service/service.go:35]: I should use `Save() ` :-)

File: domain/cloud/state/state.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (wallyworld) [reply domain/cloud/state/state.go:432]: yup

**Review by manadart — COMMENTED** [review]
One thing I missed in the initial review. We *must always* use the context receiving versions of methods, and the context must be threaded *everywhere*.

So `QueryContext` over `Query` and `ExecContext` over `Exec`.

**Comment by wallyworld** [comment]
> One thing I missed in the initial review. We _must always_ use the context receiving versions of methods, and the context must be threaded _everywhere_.
> 
> So `QueryContext` over `Query` and `ExecContext` over `Exec`.

I thought I did them all but missed some, fixed.

File: domain/cloud/state/state.go [hunk 5]
```diff
@@ -0,0 +1,441 @@
...
+			return errors.Trace(err)
+		}
+	}
+	return nil
+}
+
+func updateCACerts(ctx context.Context, tx *sql.Tx, cloudUUID string, certs []string) error {
+	// Delete any existing ca certs - we just delete them all rather
+	// than keeping existing ones as the cert values are long strings.
+	q := `
```
[the remark this answers is not in the document]
Comment (manadart) [reply domain/cloud/state/state.go:367]: I think the `WatchCloudSpecChanges` functionality will not be replicated accurately with this behaviour.

**Review by manadart — APPROVED** [review]
I still think we'll need refinement here, but we can land and iterate.

**Pull request merged**